src/heap.rs

```diff
@@ -17,9 +20,42 @@ impl HeapId {
     }
 }
 
-#[derive(Debug, Clone, PartialEq)]
+#[derive(Clone)]
 pub struct Heap(Vec<u8>);
 
+impl fmt::Debug for Heap {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        const MAX_DEBUGGED_LEN: usize = 256;
+
+        if self.0.len() <= MAX_DEBUGGED_LEN {
+            formatter.debug_tuple("Heap").field(&self.0).finish()
+        } else {
+            formatter
+                .debug_struct("Heap")
+                .field("len", &self.0.len())
+                .field("start", &&self.0[..MAX_DEBUGGED_LEN])
+                .field("hash", &hash_for_debugging(&self.0))
+                .finish_non_exhaustive()
+        }
+    }
+}
+
+// Two heaps are considered equal even if one of them has greater size allocated (provided that all additional bytes
+// are zeroed). This is to account for the case when a bootloader heap is rolled back; heap is never shrunk.
+impl PartialEq for Heap {
+    fn eq(&self, other: &Self) -> bool {
+        let mut shorter_bytes = &self.0;
+        let mut longer_bytes = &other.0;
+        if shorter_bytes.len() > longer_bytes.len() {
+            mem::swap(&mut shorter_bytes, &mut longer_bytes);
+        }
+        *shorter_bytes == longer_bytes[..shorter_bytes.len()]
+            && longer_bytes[shorter_bytes.len()..]
+                .iter()
+                .all(|&byte| byte == 0)
+    }
+}
```

Comment:
This is ok but I found it unsatisfying. WDYT about this?

```
impl PartialEq for Heap {
    fn eq(&self, other: &Self) -> bool {
        let mut to_compare = self.0.iter().zip(&other.0);
        let compared_len = to_compare.len();

        to_compare.all(|(a, b)| a == b)
            && self.0[compared_len..].iter().all(|&byte| byte == 0)
            && other.0[compared_len..].iter().all(|&byte| byte == 0)
    }
}
```

Reply:
There are specialized implementations for comparing slices in the `core` ([this one](https://github.com/rust-lang/rust/blob/83dcdb3a5dad0ed1e3e1fadc848d3f7727b41aa5/library/core/src/slice/cmp.rs#L75) is probably relevant for this case), so I'm inclined to politely disagree with a suggestion – zipping and comparing is worse in terms of performance. Subjectively, it's less readable as well. Note that [another, more general impl](https://github.com/rust-lang/rust/blob/83dcdb3a5dad0ed1e3e1fadc848d3f7727b41aa5/library/core/src/slice/cmp.rs#L61) doesn't use zipping for performance reasons 🙃

Reply:
I actually compared the resulting assembly. It is certainly different. Mine has a lot more branching but seems to be shorter an use less stack.

Anyway, we shouldn't be discussing performance as this is only used in tests. I just find this kind of comparison always unappealing, so I tried to find an alternative.

Reply:
Being only used in tests is not a part of `PartialEq` contract, and I'd argue that with the amount of data in the bootloader heap, performance is somewhat of a concern even for tests 🙃 If you insist, I can apply your suggestion, but I really don't understand what's so unappealing about comparing slices.

Reply:
Comparing slices is good. Swapping the inputs with mem::swap I don't like but Rust doesn't have .min_max(), so not much that can be done about it.